packages/taiko-client/driver/chain_syncer/blob/blocks_inserter/pacaya.go

```diff
@@ -0,0 +1,381 @@
+					if err != nil {
+						return fmt.Errorf("failed to fetch last batch (%d): %w", meta.GetBatchID().Uint64()-1, err)
+					}
+					parentNumber = new(big.Int).SetUint64(lastBatch.LastBlockId)
+				}
+			} else {
+				parentNumber = new(big.Int).SetUint64(lastPayloadData.Number)
+			}
+
+			parent, err = i.rpc.L2ParentByCurrentBlockID(ctx, new(big.Int).Add(parentNumber, common.Big1))
+		}
+		if err != nil {
+			return fmt.Errorf("failed to fetch L2 parent block: %w", err)
+		}
+
+		log.Debug(
+			"Parent block",
+			"blockID", parent.Number,
+			"hash", parent.Hash(),
+			"beaconSyncTriggered", i.progressTracker.Triggered(),
+		)
+
+		blockID := new(big.Int).SetUint64(parent.Number.Uint64() + 1)
+		difficulty, err := encoding.CalculatePacayaDifficulty(blockID)
+		if err != nil {
+			return fmt.Errorf("failed to calculate difficulty: %w", err)
+		}
+		timestamp := meta.GetLastBlockTimestamp()
+		for i := len(meta.GetBlocks()) - 1; i >= 0; i-- {
+			timestamp = timestamp - uint64(meta.GetBlocks()[i].TimeShift)
+		}
+
+		baseFee, err := i.rpc.CalculateBaseFee(
+			ctx,
+			parent,
+			new(big.Int).SetUint64(meta.GetAnchorBlockID()),
+			true,
+			meta.GetBaseFeeConfig(),
+			timestamp,
+		)
+		if err != nil {
+			return err
+		}
+
+		log.Info(
+			"L2 baseFee",
+			"blockID", blockID,
+			"baseFee", utils.WeiToGWei(baseFee),
+			"parentGasUsed", parent.GasUsed,
+			"batchID", meta.GetBatchID(),
+			"indexInBatch", j,
+		)
+
+		// Assemble a TaikoAnchor.anchorV3 transaction
+		anchorBlockHeader, err := i.rpc.L1.HeaderByHash(ctx, meta.GetAnchorBlockHash())
+		if err != nil {
+			return fmt.Errorf("failed to fetch anchor block: %w", err)
+		}
+		anchorTx, err := i.anchorConstructor.AssembleAnchorV3Tx(
+			ctx,
+			new(big.Int).SetUint64(meta.GetAnchorBlockID()),
+			anchorBlockHeader.Root,
+			meta.GetAnchorInput(),
+			parent.GasUsed,
+			meta.GetBaseFeeConfig(),
+			meta.GetSignalSlots(),
+			new(big.Int).Add(parent.Number, common.Big1),
+			baseFee,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to create TaikoAnchor.anchorV3 transaction: %w", err)
+		}
+
+		// Decompress the transactions list and try to insert a new head block to L2 EE.
+		if lastPayloadData, err = createPayloadAndSetHead(
+			ctx,
+			i.rpc,
+			&createPayloadAndSetHeadMetaData{
+				createExecutionPayloadsMetaData: &createExecutionPayloadsMetaData{
+					BlockID:               blockID,
+					ExtraData:             meta.GetExtraData(),
+					SuggestedFeeRecipient: meta.GetCoinbase(),
+					GasLimit:              uint64(meta.GetGasLimit()),
+					Difficulty:            common.BytesToHash(difficulty),
+					Timestamp:             timestamp,
+					ParentHash:            parent.Hash(),
+					L1Origin: &rawdb.L1Origin{
+						BlockID:       blockID,
+						L2BlockHash:   common.Hash{}, // Will be set by taiko-geth.
+						L1BlockHeight: meta.GetRawBlockHeight(),
+						L1BlockHash:   meta.GetRawBlockHash(),
```

Comment:
```suggestion
						L1BlockHeight: nil,
						L1BlockHash:   common.Hash{},
```

Reply:
Due to this [IsPreconfBlock check](https://github.com/taikoxyz/taiko-geth/pull/367#discussion_r1935676468) in taiko-geth.